packages/collection/src/collection.ts

```diff
@@ -626,7 +626,7 @@ export class Collection<Key, Value> extends Map<Key, Value> {
 		const iterator = this.entries();
 		if (initialValue === undefined) {
 			if (this.size === 0) throw new TypeError('Reduce of empty collection with no initial value');
-			accumulator = iterator.next().value[1];
+			accumulator = iterator.next().value![1];
```

Comment:
```suggestion
			accumulator = iterator.next().value![1] as unknown as InitialValue;
```

As per #10260. This is forced to overcome an assignability error once the `value` type is no longer `any`.

Reply:
I did look at #10260 and I don't really follow why `!`isn't okay here. Could you elaborate?

Reply:
Apologies.

This becomes directly equivalent to [the corresponding statement](https://github.com/discordjs/discord.js/blob/2fcf1910ee30935d27624a00fa2330b4983d4b43/packages/collection/src/collection.ts#L659) in `reduceRight()`, where this coercion is already used. Essentially, the value of `iterator.next().value![1]` becomes `Value`, not `InitialValue`, and TypeScript can't infer that these types should be the same at this point. ([TSP](https://www.typescriptlang.org/play/?ts=5.6.0-dev.20240724#code/MYGwhgzhAEDCD2IQFNgBcCW8B2AeA0sgJ4A00AamCAK7IB80yAHmstgCYwCyYADgcTKUa9aAG8AUAEhe1AEYgMwaACdk7asGS4AktgyYqw2tAC8FKrToAKaVIBm2AFzRrYYMGoBbauDTwVFz0DDCNLZDIAN3CXYwjoAGtiF0JSaGBEFHQsZ2g0AAsMCABKMwZgwxA4kjsMfUq4gH4g+tCq8JqpYpaQsJFxWvtXNCJeZHghx2gAQlNzAHJ7amxsnHnSgpV4AHdobGRdgBVR5ABRFS2VawADABIxRwBfaCK9+DRoMGgllcwca+KAG47CgPu5PD4-AFpj0GuFgXYMtgIB8DMgVGB-CozHlChAAHRsNAqDDICDWIGDVx1XrtfpzczLdjIex1dSlSRSLkYIbWApFfEQDAAL2QZjm0AADBt8ltdvsjidzpdrPMAErqTRiiaMLy8EbpTKoP7YaDbAz5N4vVpUaDRETrYFcqTg7y+TEBHFojFY-H7FgU-H22jTADaAEYALpOqTPZAgCBizlc12Qj3Y8w0uEiGOPCR2eye6xIlHQUNJNLB5CR6A673pjl2F0eN1QjPfbBuFtprFRDqJQS4oqUrl5uxqNDUFSm1PurFOvN5oA))

Reply:
Really wish we could constrain InitialValue based on the passed parameters (as TS correctly does for the callback). I don't love this solution, but I can't think of a "more correct" solution off the top of my head. Might have something to do with the type of the accumulator or something like that.